crates/api/src/config.rs

```diff
@@ -0,0 +1,176 @@
+//! Types for use when configuring kitsune2 modules.
+
+/// helper transcode function
+fn tc<S: serde::Serialize, D: serde::de::DeserializeOwned>(
+    s: &S,
+) -> std::io::Result<D> {
+    Ok(serde_json::from_str(&serde_json::to_string(s)?)?)
+}
+
+/// Denotes a type used to configure a specific kitsune2 module.
+///
+/// Note, the types defined in this struct are specifically for configuration
+/// that cannot be changed at runtime, the likes of which might be found
+/// in a configuration file.
+///
+/// If a specific module has a config that can be changed at runtime, the
+/// component found in this type might be a `default_` prefixed version
+/// of it, then the runtime value can be altered through different means.
+///
+/// It is highly recommended that you expose this struct in your module
+/// docs to help devs using your module understand how to configure it.
+pub trait ModConfig:
+    'static
+    + Sized
+    + Default
+    + std::fmt::Debug
+    + serde::Serialize
+    + serde::de::DeserializeOwned
+    + Send
+    + Sync
+{
+}
+
+/// Kitsune configuration.
+#[derive(Debug, Default, serde::Serialize, serde::Deserialize)]
+pub struct Config(serde_json::Map<String, serde_json::Value>);
+
+impl Config {
+    /// When kitsune2 is generating a default or example configuration
+    /// file, it will pass a mutable reference of this config struct to
+    /// the module factories that are configured to be used. Those factories
+    /// should call this function any number of times to add any default
+    /// configuration parameters to that file.
+    pub fn add_default_module_config<M: ModConfig>(
+        &mut self,
+        module_name: String,
+    ) -> std::io::Result<()> {
+        if self.0.contains_key(&module_name) {
+            return Err(std::io::Error::other(format!(
+                "Refusing to overwrite conflicting module name: {module_name}"
+            )));
+        }
+        self.0.insert(module_name, tc(&M::default())?);
+        Ok(())
+    }
+
+    /// When kitsune2 is initializing, it will call the factory function
+    /// for all of its modules with an immutable reference to this config
+    /// struct. Each of those modules may choose to call this function
+    /// to extract a module config. Note that this config is loaded from
+    /// disk and can be edited by humans, so the serialization on the module
+    /// config should be tolerant to missing properties, setting sane defaults.
+    /// A module may choose to warn about missing or extraneous properties.
```

Comment:
```suggestion
    /// A module may choose to warn about missing properties and should warn about extraneous properties.
```

Reply:
I really dislike it when I can't spell and the code doesn't tell me I gave it a property it doesn't know about :)